var/spack/repos/builtin/packages/py-submitit/package.py

```diff
@@ -0,0 +1,21 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+
+
+class PySubmitit(PythonPackage):
+    """Python toolbox for submitting jobs to Slurm."""
+
+    homepage = "https://github.com/facebookincubator/submitit"
+    pypi = "submitit/submitit-1.3.3.tar.gz"
+
+    version("1.3.3", sha256="efaa77b2df9ea9ee02545478cbfc377853ddf8016bff59df6988bebcf51ffa7e")
+
+    depends_on("python@3.6:", type=("build", "run"))
+    depends_on("py-setuptools", type=("build", "run"))
+
+    depends_on("py-cloudpickle@1.2.1:", type="run")
+    depends_on("py-typing-extensions@3.7.4.2:", type="run")
```

Comment:
These are also needed at build-time

Reply:
These packages are defined in `install_requires` and they are run-time dependencies, and the spack build doesn't fail even when type='build' is not specified (`python setup.py build` would succeed even without them).
Could you please explain why they should be defined also for build?

Reply:
My understanding of `install_requires` is that `setuptools` will check to see if these packages are already installed and if not will use `pip` to download and install them. We want to avoid that by making sure they are in `PYTHONPATH` during build-time.

Reply:
I'm pretty sure that `setuptools` doesn't consider `install_requires` for the build process.
The requirements to build the package were once specified in `setup_requires` in `setup.py`, but it's now deprecated in favour of `requires` in `pyproject.toml`, as defined in PEP-518 Specifying Minimum Build System Requirements for Python Projects.
I also tried to build this package using a clean virtualenv (containing only the most up to date versions of `pip` and `setuptools`), and it succeeded without installing additional packages.

Reply:
According to https://packaging.python.org/guides/distributing-packages-using-setuptools/#install-requires pip does install them. Will do some testing on my end as well.

Reply:
Thanks, it sounds good if you can test it as well.
From the first link you posted:
> When the project is installed by pip, this is the specification that is used to install its dependencies

And from this other link https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#declaring-required-dependency:
> setuptools support automatically download and install these dependencies when the package is installed

But both the quoted paragraphs describe the "installation", not the "building" the package.
For building, this section could be useful: https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html#package-requirement

Reply:
Sorry, terms like "build" and "install" often get conflated with many meaning.

In Spack, `type='build'` dependencies are added to the `PYTHONPATH` when you run `spack install`, and `type='run'` dependencies are added when you run `spack load` or `module load`. When we install Python packages, we run both `python setup.py build` and `python setup.py install`. So in order for the install not to add those deps, I think they need to be type=build/run, right?

Reply:
Thanks for the useful clarification (I was considering `type=build -> python setup.py build`, `type=run -> python setup.py install`).

However, I still didn't find any evidence of issues when these two dependencies (py-typing-extensions and py-cloudpickle) are defined with `type='run'` only.

Using that configuration, I executed `spack install -v py-submitit` and checked the result:
all the required packages (including py-typing-extensions and py-cloudpickle) are correctly built (with `python setup.py build`) and installed (with `python setup.py install`) before py-submitit.

It seems that `spack install` builds the dependencies considering both the `build` and `run` types, and it's confirmed by this debug log printed after executing the install command:
```
Processing dependencies for py-submitit-1.3.3-py4oeapudwwqpbypkxvdq4q63pokftui: ('build', 'link', 'run')
```

The deptypes are returned by `get_deptypes`, and they always include `link` and `run`:
https://github.com/spack/spack/blob/f5ab3ad82a62585589296ff37ebf7af1aed8303f/lib/spack/spack/installer.py#L2172:L2188

So at the end, the dependencies with `type=('build', 'run')` and `type='run'` have the same effect for `spack install`, because `run` dependencies are always installed.
Does it make sense, and is this the expected behaviour?
In this case, do you still suggest to specify both the dependency types?
Does it mean that every python dependency at runtime should always be specified with `type=('build', 'run')`?

Reply:
> However, I still didn't find any evidence of issues when these two dependencies (py-typing-extensions and py-cloudpickle) are defined with type='run' only.

Just looked into this. I think the reason that you don't see issues is that we are adding these deps to the `PYTHONPATH` even when they aren't build deps. I personally think this is a bug and we should only add deps to the `PYTHONPATH` if they are build deps.

> Does it mean that every python dependency at runtime should always be specified with type=('build', 'run')?

Yes, this is documented here: https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#python-dependencies